File: besttimetobuyandsellstock.java
/*
Problem: Best time to buy and sell stock
Language: java
Time complexity:O(n)
Space complexity:O(1)
*/
public class besttimetobuyandsellstock {
    public int maxProfit(int[] prices) {
        int max=prices[0],min=prices[0], fi=0;
        for(int i=1;i<prices.length;i++){
            if(prices[i]<min){
                if(max-min>fi)
                    fi=max-min;
                min=prices[i];
                max=0;

            }
            if(prices[i]>max){
                max=prices[i];
            }
        } 
        if(max-min>fi)
            fi=max-min;  
        return fi;
    }
    /*
     * Updated solution 
     */
    public int maxProfit(int[] prices) {
        int min=Integer.MAX_VALUE, profit=0;
        for(int i=1;i<prices.length;i++){
            if(prices[i]<min){
                min=prices[i];
            }
            else if (prices[i]-min>profit){
               profit=prices[i]-min;
            }
        } 
         
        return profit;
    }
}

File: JavaScript/bestTimetoBuyandSellStock.js
var maxProfit = function(prices) {
    let min=prices[0],maxProfit=0;
    for(let index=1;index<prices.length;index++){
        if(prices[index]>=min){
            maxProfit=Math.max(maxProfit,prices[index]-min)
        }
        if(prices[index]<min){
            min=prices[index]
        }
    }
    return maxProfit
};